phoenix-core-server/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -1642,7 +1701,14 @@ private List<Mutation> generateOnDupMutations(BatchMutateContext context, Put at
                   transferAttributes(atomicPut, put);
                   mutations.add(put);
               }
-              put.add(cell);
+              if (checkCellNeedUpdate(cell, currColumnCellExprMap)) {
+                  put.add(cell);
+              }
+
+              if (checkCellNeedUpdate(cell, currColumnCellExprMap)) {
```

Comment:
You are checking twice. Seems like a typo

Reply:
@jinggou Will doing like this simplify your logic
```suggestion
              for (int i = 0; i < tuple.size(); i++) {
                   Cell cell = tuple.getValue(i);
                   if (cell.getType() == Cell.Type.Put && checkCellNeedUpdate(cell, currColumnCellExprMap)) {
                     if (put == null) {
                       put = new Put(rowKey);
                       transferAttributes(atomicPut, put);
                       mutations.add(put);
                    }
                    put.add(cell);
```

Reply:
This makes sense. But if the put is null (no cells need update) and delete is not null, we still need to add the empty cell to put. I'm not sure about the specific logic of delete, but if we don't add an empty cell, NullPointerException will be thrown in the delete test. 
I'll update the code based on your suggested change, and we can see which way is simpler.